src/esp32/single_encoder.rs

```diff
@@ -0,0 +1,341 @@
+        Ok(tot)
+    }
+    pub fn setup_pcnt(&mut self) -> anyhow::Result<()> {
+        unsafe {
+            match esp_idf_sys::pcnt_unit_config(&self.config as *const pcnt_config_t) {
+                ESP_OK => {}
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+        }
+
+        unsafe {
+            match esp_idf_sys::pcnt_counter_pause(self.config.unit) {
+                ESP_OK => {}
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+            match esp_idf_sys::pcnt_counter_clear(self.config.unit) {
+                ESP_OK => {}
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+        }
+
+        isr_install()?;
+
+        match self.dir {
+            Direction::Forwards | Direction::StoppedForwards => {
+                esp!(unsafe {
+                    esp_idf_sys::pcnt_isr_handler_add(
+                        self.config.unit,
+                        Some(Self::irq_handler_increment),
+                        self.pulse_counter.as_mut() as *mut PulseStorage as *mut _,
+                    )
+                })?;
+            }
+            Direction::Backwards | Direction::StoppedBackwards => {
+                esp!(unsafe {
+                    esp_idf_sys::pcnt_isr_handler_add(
+                        self.config.unit,
+                        Some(Self::irq_handler_decrement),
+                        self.pulse_counter.as_mut() as *mut PulseStorage as *mut _,
+                    )
+                })?;
+            }
+        };
+
+        unsafe {
+            match esp_idf_sys::pcnt_set_filter_value(self.config.unit, MAX_GLITCH_MICROSEC * 80) {
+                ESP_OK => {}
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+            match esp_idf_sys::pcnt_filter_enable(self.config.unit) {
+                ESP_OK => {}
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+        }
+
+        unsafe {
+            match esp_idf_sys::pcnt_event_enable(self.config.unit, pcnt_evt_h_lim) {
+                ESP_OK => {}
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+            match esp_idf_sys::pcnt_event_enable(self.config.unit, pcnt_evt_l_lim) {
+                ESP_OK => {}
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+        }
+
+        Ok(())
+    }
+
+    #[inline(always)]
+    #[link_section = ".iram1.pcnt_srv"]
+    unsafe extern "C" fn irq_handler_increment(arg: *mut core::ffi::c_void) {
+        let arg: &mut PulseStorage = &mut *(arg as *mut _);
+        let mut status = 0;
+        esp_idf_sys::pcnt_get_event_status(arg.unit, &mut status as *mut c_ulong);
+        if status & pcnt_evt_h_lim != 0 {
+            arg.acc.fetch_add(1, Ordering::SeqCst);
+        }
+        if status & pcnt_evt_l_lim != 0 {
```

Comment:
do we need both?

Reply:
Probably not? I recall trying using only one or the other and finding it not to work, but there may have been other problems with my code at the time. I'll try removing